### ISSUE TITLE
const overload for in CascadeStages::getCascadeStorage()

### ISSUE THREAD

**Issue by stellarpower** [berndporr/iir1#66]
[`getCascadeStorage()`](https://github.com/berndporr/iir1/blob/master/iir/Cascade.h#L167 ) returns a const struct to the held co-efficients (which itself is a pointer inside, so that is mutable from the outside).

Would it be appropriate to add a const overload for this function? I'm iterating over the co-efficients from the outside without modifying them, so my filter is taken by const reference. 

Personally I'd usually probably keep it as a subobject and return a const/mutable reference to the data directly inside, rather than make a copy of Cascade::Storage and then return this by value - but I assume there may well be reasons why it makes more sense as it currently is. As the issue here would be in order to enforce not modifying the storage, Cascade::Storage would then need to propagate the const to the pointer itself.

Thanks

**Comment by berndporr** [comment]
Thanks for pointing this out. I've tried it just by adding const but because the coeffs are public one could still manipulate them. So now I've done a proper copy operation to a temp structure. I don't think you'll be using this often so that the additional execution time shouldn't matter. Feel free to try it out.

**Comment by berndporr** [comment]
Unit tests failed with the copy operation as internally the filter manipulates the coeffs via this pointer. I guess we just leave it as it is as perhaps users are actually use this also to manipulate coeffs.

**Comment by nickbailey** [comment]
Great testing!

I guess it involves a refactoring at some point. If I get any spare time I'll give you a hand! It'd be a pleasure.

**Comment by stellarpower** [comment]
Thanks, yes, think there potentially could be a few refactorings. Not really looked that much at the internals but using a few filters regularly.

In particular the SOS with specified co-efficisnts; like how it's templated on the order as that's efficient, but at the same time have written a few wrapper classes and had to pre-instantiate without there being a common parent class for various filters provided. So maybe some further changes would help in general.

Similarly, if I end up needing to work morebsoth these, would happily have a refactor and open an MR.

Also not sure if it's interesting to you, but have been working the last year on neural nets to perform faster filtration than conventional algorithms. It still needs some work (probably custom loss function, it didn't learn until using DTW and that's largely time-invariant), but current implementation can provide a (wobbly) but smooth estimate of the bandpower for low frequencies with less group delay than an equivalent elliptical filter has kicked in. It sort of sits in-between a filter and an FFT, being closer to a non-negative band amplitude than a filter but lower latency than either that or an FFT.  Trained against a constant-Q; time permitting hopefully will publish it some day. But could link some plots.
